ruby/sdk-phase-1.md

```diff
@@ -0,0 +1,439 @@
+    be instance methods (since we would allow class methods as activities).
+* Activity classes are instantiated at registration time, not for each invocation.
+  * 💭 Why? Existing Temporal Ruby SDK does instantiate for each activity task, but this discourages use of shared
+    objects or state to be referenced. This does have the tradeoff our other SDKs have of requiring the function to be
+    thread safe.
+* `Temporalio::Activity::Context` is a thread/fiber-local context available to activities.
+  * It has instance methods for info, heartbeat, etc.
+  * It has a class method for `current` that returns from thread/fiber-local and class method shortcuts for all instance
+    methods.
+    * 💭 Why class methods equivalents? `Temporalio::Activity::Context.heartbeat` is easier to read than
+      `Temporalio::Activity::Context.current.heartbeat`. In .NET we made them get current instance first because that's
+      normal there, and in Python we only offered package-level methods because that's normal there.
+    * 💭 Why allow access to an instance? Some users want to pass the activity context to another thread.
+  * 💭 Why `Temporalio::Activity::Context` instead of just `Temporalio::Activity` or putting helpers on base class or
+    something? It is a common requirement in our SDKs to be able to access the activity context from the environment.
+    Also, it is a common requirement to be able to do that from utility methods outside of the explicit activity call.
+* Activity can choose its executor, default is `:default`.
+  * See worker section for more information on executors.
+* Cancellation is reported to the executor.
+  * For thread pool executor (the default), this raises a cancellation exception via `Thread.raise`.
+  * For fiber executor, this is done via `Fiber.raise`.
+  * A method on the activity context, `shield`, can be given a block that will hold off on cancelling until after it is
+    done. Nested `shield` calls only raise after the last one is complete.
+* Type-safe invocation from workflows is TBD.
+
+## Workers
+
+```ruby
+module Temporalio
+  class Worker
+    # Run multiple workers until signal reached, block optional
+    def self.run(*workers, shutdown_signals: [], &block)
+      ...
+    end
+
+    def initialize(
+      client,
+      task_queue:,
+      activities: [],
+      activity_executors: nil,
+      ...,
+    )
+      ...
+    end
+
+    # Run until shutdown called or block completed (which calls shutdown)
+    def run(&block)
+      ...
+    end
+
+    # Shutdown worker
+    def shutdown()
+      ...
+    end
+  end
+end
+```
+
+* Following the API of the current Temporal Ruby SDK, workers can be run individually, but we will also provide a class
+  method to run several at once.
+  * 💭 Why is something needed to run multiple workers for the user? Because in Ruby they don't have easy ways of
+    running multiple blocking things at once without external libraries or forcing threads on users.
```

Comment:
In most of our SDK languages there are better ways to run multiple workers at once without eating a thread per worker. JS is always single threaded, Python has `asyncio.gather`, Go has goroutines, and .NET has tasks. In Java we have a `WorkerFactory` that lets you run multiple workers at once because it lacks good language-level reuse-thread capabilities. Ruby has the same problem. So I was thinking a `WorkerPool` or similar that can run many workers in a single blocking run call.